### PR TITLE
[HM3D]--Connected Components Utility; CC-based Semantic BBox generation; Mesh connectivity reporting

### PR DESCRIPTION
**Pull request by jturner65**
## Motivation and Context
This PR adds the following functionality : 
- Derive connected components on a set of vertices conditioned on some vertex attribute, such as color.  It finds all cc's that share the same vertex attribute value, finds the AABB of the resultant vertex cloud, and aggregates the results into a map keyed by semantic object IDX.

- Semantic BBoxes are being generated when needed (currently only in HM3D dataset) using CC with largest BBox (by volume) for a particular semantic annotation.

 - Reporting tools are provided through viewer, including the ability to go through all the scenes in a dataset to derive and save the report to csv file.  A python binding is also provided to give access to this data.

## How Has This Been Tested
All existing local c++ and python tests pass. 

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/esp/assets/GenericSemanticMeshData.cpp [hunk 1]
```diff
@@ -327,11 +331,118 @@ GenericSemanticMeshData::partitionSemanticMeshData(
...
+      geo::conditionalDFS(adjList, cpu_cbo_, vIDX, visited, vertColor,
+                          setOfVerts);
+      const uint32_t colorKey = getColorAsInt(vertColor);
+      // build map keyed by color of vert set for cc
+      auto findIter = clrsToComponents.find(colorKey);
+      if (findIter == clrsToComponents.end()) {
+        // not found already
+        findIter = clrsToComponents.insert({colorKey, {}}).first;
+      }
+      findIter->second.push_back(setOfVerts);
```
Comment (Skylion007) [inline src/esp/assets/GenericSemanticMeshData.cpp:415]:
```suggestion
      findIter->second.emplace_back(std::move(setOfVerts));
```

File: src/esp/assets/GenericSemanticMeshData.cpp [hunk 2]
```diff
@@ -327,11 +331,118 @@ GenericSemanticMeshData::partitionSemanticMeshData(
...
+  for (const auto& elem : clrsToComponents) {
+    const uint32_t colorKey = elem.first;
+    const std::vector<std::set<uint32_t>>& vectorOfSets = elem.second;
+    auto findIter = results.find(colorKey);
+    if (findIter == results.end()) {
+      // not found already
+      findIter = results.insert({colorKey, {}}).first;
+    }
+    for (const std::set<uint32_t>& vertSet : vectorOfSets) {
+      findIter->second.push_back(
```
Comment (Skylion007) [inline src/esp/assets/GenericSemanticMeshData.cpp:438]:
```suggestion
      findIter->second.emplace_back(
```

File: src/esp/assets/ResourceManager.cpp [hunk 3]
```diff
@@ -215,6 +217,54 @@ void ResourceManager::initPhysicsManager(
...
+  }
+  // build map with key being semanticObject IDX in semantic Objects array
+  std::unordered_map<uint32_t, std::vector<std::pair<int, esp::geo::OBB>>>
+      results;
+  for (auto& elem : mapColorIntsToSemanticObjIDXs) {
+    const auto objID = elem.second;
+    const auto colorAsInt = elem.first;
+    auto mapEntry = semanticRes.find(colorAsInt);
+    if (mapEntry != semanticRes.end()) {
+      results.insert({objID, std::move(mapEntry->second)});
```
Comment (Skylion007) [inline src/esp/assets/ResourceManager.cpp:261]:
```suggestion
      results.emplace(objID, std::move(mapEntry->second));
```

File: src/utils/viewer/viewer.cpp [hunk 4]
```diff
@@ -1133,6 +1143,64 @@ bool loadTransformFromFile(const std::string& filename,
...
+    const auto clr = baseObj->getColor();
+    for (const std::pair<int, esp::geo::OBB>& elem : listOfObbs) {
+      const auto& obb = elem.second;
+      const std::string dataString = Cr::Utility::formatString(
+          "{},{},{} {} {},{},{} {} {}, {} {} {}", objIDX, baseObj->id(),
+          clr.r(), clr.g(), clr.b(), elem.first, obb.center().x(),
+          obb.center().y(), obb.center().z(), obb.sizes().x(), obb.sizes().y(),
+          obb.sizes().z());
+      ESP_DEBUG() << dataString;
+      file << dataString << "\n";
```
Comment (Skylion007) [inline src/utils/viewer/viewer.cpp:1197]:
```suggestion
      file << dataString << '\n';
```

File: src/esp/assets/ResourceManager.cpp [hunk 5]
```diff
@@ -215,6 +217,54 @@ void ResourceManager::initPhysicsManager(
...
+  GenericSemanticMeshData::uptr semanticMeshData =
+      flattenImportedMeshAndBuildSemantic(*fileImporter_, semanticInfo);
+
+  // return connectivity query results - per color map of vectors of pairs of
+  // vert counts and OBBs
+  auto semanticRes = semanticMeshData->buildSemanticCCReportData();
+  // build temp map of colors to SemanticObject IDs
+  const auto& semanticObjs = semanticScene_->objects();
+
+  std::unordered_map<uint32_t, uint32_t> mapColorIntsToSemanticObjIDXs;
```
Comment (Skylion007) [inline src/esp/assets/ResourceManager.cpp:248]:
```suggestion
  std::unordered_map<uint32_t, uint32_t> mapColorIntsToSemanticObjIDXs;
  mapColorIntsToSemanticObjIDXs.reserve(mapColorIntsToSemanticObjIDXs.size());
```

File: src/esp/assets/GenericSemanticMeshData.cpp [hunk 6]
```diff
@@ -327,11 +331,118 @@ GenericSemanticMeshData::partitionSemanticMeshData(
...
+  // 1 entry per color
+  std::unordered_map<uint32_t, std::vector<std::pair<int, esp::geo::OBB>>>
+      results(clrsToComponents.size());
+  for (const auto& elem : clrsToComponents) {
+    const uint32_t colorKey = elem.first;
+    const std::vector<std::set<uint32_t>>& vectorOfSets = elem.second;
+    auto findIter = results.find(colorKey);
+    if (findIter == results.end()) {
+      // not found already
+      findIter = results.insert({colorKey, {}}).first;
```
Comment (Skylion007) [inline src/esp/assets/GenericSemanticMeshData.cpp:435]:
```suggestion
      findIter = results.emplace(colorKey, {}).first;
```

**Review by Skylion007 — DISMISSED** [review]
Performance improvements

**Comment by jturner65** [comment]
Thanks as always @Skylion007 :)

File: src/esp/assets/ResourceManager.cpp [hunk 7]
```diff
@@ -307,14 +358,17 @@ void ResourceManager::buildSemanticColorMap() {
 
 void ResourceManager::buildSemanticColorAsIntMap() {
   semanticColorAsInt_.clear();
+  if (semanticColorMapBeingUsed_.size() == 0) {
```
Comment (Skylion007) [inline src/esp/assets/ResourceManager.cpp:361]:
```suggestion
  if (semanticColorMapBeingUsed_.empty()) {
```

File: src/esp/geo/Geo.h [hunk 8]
```diff
@@ -5,8 +5,12 @@
 #ifndef ESP_GEO_GEO_H_
 #define ESP_GEO_GEO_H_
 
+#include <set>
+#include <typeinfo>
```
Comment (Skylion007) [inline src/esp/geo/Geo.h:9]: Maybe I missed it, but it this include necessary?

Comment (jturner65) [reply src/esp/geo/Geo.h:9]: yep.  using both sets for cc's and typeinfo for error message.

**Pull request merged**